libsolidity/analysis/ContractLevelChecker.cpp

```diff
@@ -510,3 +630,246 @@ void ContractLevelChecker::checkBaseABICompatibility(ContractDefinition const& _
+	{
+		auto [begin,end] = collectedFunctions.equal_range(*it);
+
+		// Only one function
+		if (next(begin) == end)
+			continue;
+
+		// Not an overridable function
+		if (!(*it)->isOverridable())
+			continue;
+
+		// Function has been explicitly overridden
+		if (contains_if(
+				contractFuncs,
+				[&] (FunctionDefinition const* _f) {
+					return hasEqualNameAndParameters(*_f, **it);
+				}
+			)
+		)
+			continue;
+
+		set<FunctionDefinition const*> ambiguousFunctions;
+		SecondarySourceLocation ssl;
+
+		for (;begin != end; begin++)
+		{
+			ambiguousFunctions.insert(*begin);
+			ssl.append("Definition here: ", (*begin)->location());
+		}
+
+		// Make sure the functions are not from the same base contract
+		if (ambiguousFunctions.size() == 1)
+			continue;
+
+		m_errorReporter.typeError(
+			_contract.location(),
+			ssl,
+			"Derived contract must override function " +
+			(*it)->name() +
+			". Function with the same name and parameter types defined in two or more base classes."
+		);
+	}
+}
+
+set<ContractDefinition const*> ContractLevelChecker::resolveOverrideList(OverrideSpecifier const& _overrides) const
+{
+	set<ContractDefinition const*> resolved;
+
+	for (ASTPointer<UserDefinedTypeName> const& override: _overrides.overrides())
+	{
+		Declaration const* decl  = override->annotation().referencedDeclaration;
+		solAssert(decl, "Expected declaration to be resolved.");
+
+		// If it's not a contract it will be caught
+		// in the reference resolver
+		if (ContractDefinition const* contract = dynamic_cast<decltype(contract)>(decl))
+			resolved.insert(contract);
+	}
+
+	return resolved;
+}
+
+
+void ContractLevelChecker::checkModifierOverrides(FunctionSet const& _funcSet, ModifierSet const& _modSet, std::vector<ModifierDefinition const*> _modifiers)
+{
+	for (ModifierDefinition const* modifier: _modifiers)
+	{
+		if (contains_if(_funcSet, MatchByName{modifier->name()}))
+			m_errorReporter.typeError(
+				modifier->location(),
+				"Override changes function to modifier."
+			);
+
+		auto [begin,end] = _modSet.equal_range(modifier);
+
+		// Skip if no modifiers found in bases
+		if (begin == end)
+			continue;
+
+		if (!modifier->overrides())
+			overrideError(*modifier, **begin, "Overriding modifier is missing 'override' specifier.");
+
+		for (; begin != end; begin++)
+			if (ModifierType(**begin) != ModifierType(*modifier))
+				m_errorReporter.typeError(
+					modifier->location(),
+					"Override changes modifier signature."
+			);
+	}
+
+}
+
+void ContractLevelChecker::checkOverrideList(FunctionSet const& _funcSet, FunctionDefinition const& _function)
+{
+	set<ContractDefinition const*> specifiedContracts =
+		_function.overrides() ?
+		resolveOverrideList(*_function.overrides()) :
+		set<ContractDefinition const*>{};
+
+	// Check for duplicates in override list
+	if (_function.overrides() && specifiedContracts.size() != _function.overrides()->overrides().size())
+	{
+		// Sort by name to find duplicate for error reporting
```

Comment:
Identical contracts can have multiple names (for example through the import statement). Equality comparison should be done through the AST id.

Reply:
I missed that completely. Fixed that now! Thanks!